Update functions inputs

RaphaelAlvez opened the pull request:
### FOR THE MOMENT ONLY ROTATE WAS UPDATED

fileUpdater and JSON editor were not updated so far


### Current situation

![Image](https://user-images.githubusercontent.com/73476656/212735989-99ec95a4-4d62-445a-a9e9-517ed6bb88cd.png)

main problems:
- widget array/ID inputs have many names (label, canvas)
- "collection" is used as input in some functions and as output in others (CANVAS vs SELECT)
- Some routines support widgets as direct targets, and some support holder content as targets.
- Some routines support collections of target widgets
- No support of collections of holders for execution on its content

### Desired final setup

![Image](https://user-images.githubusercontent.com/73476656/212739144-79bff620-c505-4c3a-a620-fc397f9ef8c4.png)

inputs:
- `target`: Function will be executed directly to/in relation to the widgets listed or in collection
- `holder`: Function will be executed to widgets that are children of each holders. For seats, function will be executed on children of the seats hand and if needed to the correct owner
- `source`: Collection as input. The collection itself will be the target (SELECT) not the widgets inside of it
- `collection`: Collection as output. Resulting collection from the operation

main points discussed in meetings:
- Avoid changing behavior of functions that already accept more than one kind of input
- In some situations a simple parameter rename is needed (example: FLIP)
- MOVE will need an exception to the name convention as it needs 2 holder parameters.
- holder parameter will check for valid children will be checked by dropTarget
- function will be only executed in a target if holder is not defined
- note that even if holder may not be a normal input to a function, it should be included as a valid input. a holder input should be viewed as another form of pseudo collection
- single IDs as strings will no longer be supported. for one widget users should use ["idName"]
- string inputs will be used for collections only

examples to make things clear:
- RECALL at the moment only accepts holder parameters. therefore to recall cards we need to put the deck in a holder. target parameter will allow recalling deck a to the deck parent or to the deck itself if the deck is not in a holder
- FILP should stay exactly the same. the behavior of priority from holder over target comes from the current setup of flip
- CANVAS will only have its property canvas renames target
- TURN will have its property source renamed to target since the seats have their property changed
- SORT and SHUFFLE should have source as input and in the case, the order of the widgets in the collection would be changed. No property in any widget would change.

ArnoldSmith86 commented:
PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-1658/pr-test (or any other room on that server)

After merging, a backup will be available at `https://beta.virtualtabletop.io/editor/PR1658-pr-test`.

RaphaelAlvez commented:
I uploaded a showcase in the test room

robartsd commented:
I think that `holder` parameter should accept widgets of any type. If the type of the widget is not "holder" then all the children of the widget are the widgets to be acted upon (other widget types don't have a dropTarget property to filter with).

Should `to` of `MOVE` be renamed `target`?

Possibly a separate PR, but I think that `MOVE` should be given `x` and `y` parameters for use if the target of the move is not an `alignChildren` holder. `MOVEXY` could be converted to `MOVE` with a target of `null`.

RaphaelAlvez commented:
> I think that `holder` parameter should accept widgets of any type. If the `type` of the widget is not "holder" then all the children of the widget are the widgets to be acted upon (other widget types don't have a dropTarget property to filter with).

I made the documentation before started working on this. After your comment on discord I changed to be as you described with the exception that I changed the .children() code to be .childrenTarget() and it has a different behavior for seats. Seats push the hand children with the correct owner.

> Should `to` of "MOVE" be renamed `target`?

I know the word target is normally related to a destination butod rather keep it for the widgets that will be directy be moved. This is a feature we don't have are the moment.
For the destination I'm thinking on using `destination`, `targetHolder` or `toHolder`
I'm really not set in any naming convention so 

I'm open to change to better words based on feedback

> Possibly a separate PR, but I think that "MOVE" should be given `x` and `y` parameters for use if the target of the move is not an `alignChildren` holder. "MOVEXY" could be converted to "MOVE" with a target of `null`.

Probably another PR. My focus is to make sure I push this with 0 changes to current behavior. Changing more stuff will make it harder to be sure I did a good job.

I might make parallel PRs since I'll be more familiar with the code

robartsd commented:
> > I think that `holder` parameter should accept widgets of any type. If the `type` of the widget is not "holder" then all the children of the widget are the widgets to be acted upon (other widget types don't have a dropTarget property to filter with).
> 
> I made the documentation before started working on this. After your comment on discord I changed to be as you described with the exception that I changed the .children() code to be .childrenTarget() and it has a different behavior for seats. Seats push the hand children with the correct owner.

So when `holder` is specified, the collection of widget acted upon are child widgets based on the widgets in the collection passed to `holder`. Any widget type works; for most widgets types it is a simple selection of child widget; for holders the child widgets are limited to those that match dropTarget; seats redirect to the widget specified has `hand` in the seat and add a further restriction on owner matching the `player` in the seat. This sounds good to me.

What happens if a seat in the collection passed to `holder` has no valid `hand`?

RaphaelAlvez commented:
> I think that `holder` parameter should accept widgets of any type. If the `type` of the widget is not "holder" then all the children of the widget are the widgets to be acted upon (other widget types don't have a dropTarget property to filter with).

I checked and you can currently put any ID in holder and it will apply to all children. Same behavior in this PR except for seats

> What happens if a seat in the collection passed to `holder` has no valid `hand`?

fixed all possible exceptions thanks for pointing it out.

flackr commented on server/fileupdater.mjs:
```diff
@@ -1,4 +1,4 @@
-export const VERSION = 12;
+export const VERSION = 50;
```
Version 50 already? Avoiding conflicts? I assume we'll change this before merging.

RaphaelAlvez replied:
yes. I expect this PR will take some time and other versions will come before it. it ill be changed.

flackr commented on client/js/widgets/seat.js:
```diff
@@ -42,6 +42,15 @@ class Seat extends Widget {
     this.updateLinkedWidgets(true);
   }
 
+  childrenTarget() {
```
Any reason not to override children() like holder.js does?

RaphaelAlvez replied:
children() is actually used in some other places and I'm afraid it might break something if I change the seats' children().
I might be wrong; this can be an unnecessary extra function, but I feel safer this way.

This is one of the things I'm most unsure about this PR but at the same time it doesn't feel too problematic

flackr commented on client/js/widgets/seat.js:
```diff
@@ -42,6 +42,15 @@ class Seat extends Widget {
+    if (widgets.get(this.get("hand"))) {
+      return widgets.get(this.get("hand")).children().filter(c=>!c.get('owner') || c.get('owner')==this.get("player"));
+    } else {
+      return {}
```
nit: I think this should return an array - that's the type we get from children() and would get from the above branch.

RaphaelAlvez replied:
Done

flackr commented on client/js/widgets/widget.js:
```diff
@@ -1370,27 +1374,35 @@ export class Widget extends StateManaged {
+                if (w.get("type")=="seat") {
+                  problems.push(`the hand of ${w.get("id")} is empty or does not exist`);
+                } else{
+                  problems.push(`${w.get("id")} is empty or does not exist`);
```
I don't think you can get here if `w` didn't exist as it would crash trying to call methods on `w`. It looks like `getCollection` would filter out missing widgets here: https://github.com/ArnoldSmith86/virtualtabletop/blob/6f396ceddd95a306d63edab5fe428e35b4fa035f/client/js/widgets/widget.js#L697

RaphaelAlvez replied:
You're correct this does not give the intended problem for invalid id of seats and holders. I tested only for invalid hands and empty holders. 

it does not freeze the client if the ID is invalid but the debug doesn't know that there was a problem

flackr commented on client/js/widgets/widget.js:
```diff
@@ -1370,27 +1374,35 @@ export class Widget extends StateManaged {
+              if (w.childrenTarget().length){
+                for(const c of w.childrenTarget().slice(0, a.count || 999999))
+                  await c.rotate(a.angle, a.mode);
+              } else {
```
I'm not sure if it should be an error if there is nothing to ROTATE in the specified holders, just like it's not an error if there is nothing to MOVE. It seems like this would generate errors for expected states sometimes.

RaphaelAlvez replied:
I added this because we get a similar error for collections. You are correct that it might generate a bunch of "empty holder" errors that have close to no value.

I'm open to remove it

robartsd commented on client/js/widgets/holder.js:
```diff
@@ -29,7 +29,7 @@ class Holder extends Widget {
+    if(children.length == 1 && children[0].get('type') == 'pile') //this needs to be improoved so that it can support holders with multiple piles. (allignChildren = false)
       children = this.childrenFilter(children[0].children(), false);
     return children;
   }
```
If we want to resolve all first level piles (player's cannot put a pile into a pile directly, but routines might be able to - the only engine code that I am aware of that resolves nested piles is for getting `_ancestor`), not just sole piles, we could use:

```js
children() {
  const children = super.children().flatMap(child=>child.get('type')=='pile'?child.children():child);
  return this.childrenFilter(children, false)
}
```
Currently piles are not resolved if they have siblings, so this might be considered a breaking change.

RaphaelAlvez replied:
`this.childrenFilter(children, false)`

This was something I did couldn't figure out but left it because it would not cause any harm and maybe there was some edge case I was missing

robartsd replied:
Yeah, I'm not sure if the accept piles argument of `Holder.childrenFilter()` is the best approach, but we certainly need the call to `compareDropTarget()` that is implemented within `Holder.childrenFilter()`.

I actually think it might be nice to implement pile resolution as an option in `Widget.children()`.

VTT-AI-Agent commented:
I have taken this PR over and rebuilt it on current `main`. The branch was ~880 commits behind and conflicting, and the in-progress ROTATE/MOVE work no longer applied, so the merge commit resets the tree to `main` and everything after it is a fresh implementation. The design is the one this PR and the project board describe; the PR body now lists what actually landed, per function.

### Answers to the review so far

> I think that `holder` parameter should accept widgets of any type. If the `type` of the widget is not "holder" then all the children of the widget are the widgets to be acted upon (other widget types don't have a dropTarget property to filter with).

That is what it does. `holder` resolves through `widget.children()`, which is already per type: a holder filters by its `dropTarget` (and unwraps a single pile), a seat returns what lies in its hand and belongs to its player, and anything else returns its children. So no new per-type code was needed, and a seat with no hand or no player simply contributes nothing.

> Should `to` of "MOVE" be renamed `target`?

It is `toHolder` now, with `fromHolder` on the other side and `target` for the widgets themselves — the naming the project board settled on. `target` had to stay free for "the widgets being moved", which is the case MOVE could not express before.

> Possibly a separate PR, but I think that "MOVE" should be given `x` and `y` parameters […] "MOVEXY" could be converted to "MOVE" with a target of `null`.

Left for a separate PR, as agreed in the thread. MOVEXY here only gets the same names as MOVE (`fromHolder`, plus a new `target`), nothing about what it does changes.

### Two decisions I had to make that are not spelled out on the board

1. **MOVEXY** follows MOVE rather than the general rule: `fromHolder` (not `holder`), so the two functions that move widgets read alike.
2. **LABEL's `label` and TIMER's `timer`** are folded into `target` (they were widget-id parameters, so they are exactly what `target` means), and **SORT/SHUFFLE** use `source`, because they reorder the collection itself rather than changing any widget — that is what the PR description asks for.

Both are easy to change if you would rather have it another way.

### FileUpdater — please decide

The PR description said the file updater still had to be done. I deliberately did **not** add a migration, and would like a decision before doing so. Rewriting existing games to the new spelling changes the state of every library game that uses these parameters, which means re-recording the expected hashes in all four `tests/testcafe/publiclibrary-*.js` files, and it leaves the prose of the ~20 `Functions - *` tutorials describing parameter names their own JSON no longer uses. Without it, old games keep working unchanged (the engine reads the old name whenever the new one is absent) and nothing about this PR can break a published game — but `collection` stays legal as an input forever.

If you want the migration in this PR, say so and I will add it together with the re-recorded hashes.

### What the editing UI looks like

The visual routine editor gained a "what is in a holder" option for every function that has a holder now, worded like the one COUNT already had, and the holder marks the `target` it replaces as ignored, so the two are mutually exclusive in the UI. A function that still carries an old spelling keeps opening: the old name is worded in place of the current one exactly while the engine falls back to it, and its chip gets the orange "!" deprecation note that the CANVAS `canvas` parameter already had. Adding a MOVE from the editor writes `{"func":"MOVE","fromHolder":null,"count":1}` and reads *"Move 1 widget from holder(s)/collection to holder(s)/collection"*.

Screenshots from a local run:

- the editor on a ROTATE that uses `holder`: https://agent.virtualtabletop.io/reports/pr/1536305652446535690/routine-editor-holder.png
- a MOVE added from the editor: https://agent.virtualtabletop.io/reports/pr/1536305652446535690/routine-editor-new-move.png
- the new tutorial, variant "target and holder": https://agent.virtualtabletop.io/reports/pr/1536305652446535690/tutorial-target-and-holder.png
- variant "MOVE": https://agent.virtualtabletop.io/reports/pr/1536305652446535690/tutorial-move.png
- its library image: https://agent.virtualtabletop.io/reports/pr/1536305652446535690/tutorial-library-image.png
- the "i" text behind a ROTATE, which now documents `target` and `holder`: https://agent.virtualtabletop.io/reports/pr/1536305652446535690/routine-editor-info.png

If the wording of any of it reads wrong to you, say which sentence and I will change it.

### Verification

`npm test` (jest, 1206 tests) is green, and locally I ran the TestCafe files most likely to notice a behavior change — `routines.js`, `publiclibrary-1..4`, `interactions.js`, `holderevents.js` — 76 tests, all passing without touching a single expected hash.

VTT-AI-Agent commented:
Thanks — the MOVE finding was real and the reproduction was exact. All six points are addressed; here is what changed, per finding.

**1. MOVE with several destinations (high) — fixed.** `let offset = 0` is back above `for(const target of collections[destinations])`, so the count is spent across the destinations one after the other again and `MOVE target count:1` into two holders deals one widget to each. I re-ran your case against `main` and this branch and they now agree (`c1 → d1`, `c2 → d2`).

**2. A leftover `from: ""` (medium) — fixed.** `if(!a.from) delete a.from;`, which restores the old truthiness rule exactly. Pinned by a test that uses `from: ''` rather than `from: null`.

**3. Temporary collections (medium) — fixed.** The routine loop remembers the counter before each operation and deletes every `$collection_N` the operation allocated once it is done, so nothing accumulates for a `FOREACH`, nothing extra gets deep-copied by a `CALL`, and `DELETE` walks the same collections it always did. Nested routines are safe: one that shares `collections` by reference takes its own snapshot after the outer temporaries exist, so it only ever removes its own.

**4. CANVAS on an empty holder (low) — fixed**, and the same guard was missing in **LABEL and TIMER**, which reported `Collection DEFAULT is empty` for an empty holder — `a.target` there is the default the operation never read. All three now only report an empty *collection*, the way FLIP and ROTATE do.

**5. Test coverage — added.** `tests/client/engine/widget-parameters.test.js` gained: dealing to several destinations with `count` (new and old spelling), the same with `fillTo`, the half-filled `from: ''`, and one that asserts no `$collection_N` survives its operation. I checked each of them fails on the code as it was before this round.

**6. Smaller notes.**
- **CALL with `return: false`** — kept as it is (it aborts the caller anyway), and the help text now says so, together with the fact that `variable`/`collection` come from the last callee when several widgets are named.
- **MOVEXY** — left as it is, but written down under "Also fixed" in the PR body so it is a visible decision. I checked all 573 library files: exactly one has a bare `{"func":"MOVEXY"}` (`library/games/Card Games`, `gameID25/dealRoutine`) and it runs with no `DEFAULT` collection in scope, so nothing published changes. Happy to make it report a missing parameter instead if you would rather have the old error.
- **FileUpdater** — agreed, still deferred.

**Tutorial.** Both notes taken: each variant names itself in its title now (`Functions: widget inputs (target and holder)` / `(MOVE)` / `(seats and counting)`), and there is a third variant for exactly the two facets you named — a `count` spent per holder against the same widgets named as one collection, and `holder` on a seat, including the MOVE into a seat that gives the widgets to its player in the first place (without which the seat contributes nothing, which the variant says out loud). Screenshots: [the room](https://agent.virtualtabletop.io/reports/pr/1536305652446535690/tutorial-seats-and-counting.png), [after pressing every button](https://agent.virtualtabletop.io/reports/pr/1536305652446535690/tutorial-seats-and-counting-in-use.png).

**Verification.** `npm test` green (1214 tests), and locally `routines.js`, `publiclibrary-1..4`, `interactions.js` and `holderevents.js` — 71 tests — pass without touching a single expected hash.

**beta label** — agreed, not added; this is an engine change.

VTT-AI-Agent commented:
Thanks — this was a useful pass; the two big ones were both real and both are now said out loud in the sentence. Everything is in `09a2f2c` on this branch. Per finding:

### 1. A `target` set alongside a `holder` vanishes — fixed, and generalised

The reason `targetReplacedBy()` computes was indeed never rendered. It is now: **every** parameter the engine skips because of how another one is set stays in the sentence, struck through and faded, with a red "!" behind it whose popup says why nothing comes of it and that the ⊖ takes it out without changing anything. Your example now reads `Click what is in h1, ~~target~~ card1 ❗`.

It is not special-cased to holder/target — the `ignored` mechanism already existed for `MOVE count` vs `fillTo`, `CANVAS x/y/color`, `MOVEXY z: 0`, `AUDIO silence` and `IF condition`, and all of them were being dropped just as silently (the header comment in `routine.js` even claimed they got a red "!" in a list view that no longer exists). One case I added while I was there: the *old* spelling of the widgets is skipped for exactly the same reason the current one is, so `{"func":"SET","holder":"h1","collection":"aces"}` now shows `collection aces` struck through too, where it used to disappear.

A parameter the operation does *not* have is still neither shown nor offered — nothing new appears on a clean operation.

![ignored](https://agent.virtualtabletop.io/reports/pr/1536305652446535690/ui-popup-ignored.png)

### 2. A `count` is spent per holder — fixed, in your wording

`Rotate 1 widget in each of h1 and h2 by 45 degrees`, and the same on `FLIP`. Also on the two sides of `MOVE`, which deals per source *and* per destination: `Move 1 widget from each of h1 and h2 to d1`, `Move 1 of the widgets called aces to each of d1 and d2`, and `… to d1 and d2 until each of them holds 3` when it is topping them up. `CANVAS` words it in its own option: `, for 1 canvas in each of them`.

"each of" only appears where the count really is spent more than once: a list of two or more, or a collection name (which may hold any number), and only while the count is a limit at all — `Turn all widgets in h1 and h2 face up` says nothing about "each", because taking everything out of both holders is the same thing said twice. A single id, or a `${...}` the editor cannot count, keeps the short wording rather than guessing at a plural.

### 3. Three ways to reach a holder — evened out

`FLIP`, `ROTATE`, `SORT` and `SHUFFLE` now offer **`what is in a holder`** in their "add option" menu like every other operation, instead of only through the combined chip (which is gone — they use the same `holderClause()` + `targetClauses()` pair as CANVAS/CLICK/DELETE now). `CALL` and `MOVEXY` lost their differently-worded labels, so the phrase is identical on all thirteen. There is a test pinning that.

`MOVE`, `MOVEXY`, `COUNT` and `RECALL` keep asking it through the **way-of-working drop-down** (`Move widgets from a holder`, `Count what is in a holder`) rather than as an option, and I left that as it is on purpose: their whole sentence reads differently with a holder — a MOVE out of a holder counts 1 by default and the picked widgets count "all" — so it is a variant, not something you switch on. The same drop-down is the first thing on the card, so it is not less discoverable than the option; it is a different question.

### 4. The deprecation popup broke into mid-sentence paragraphs — fixed

One line per sentence in `deprecatedNameNote()`, so it renders as three paragraphs that each end where the sentence does. The `skip` note had the same problem and got the same treatment.

![deprecation](https://agent.virtualtabletop.io/reports/pr/1536305652446535690/ui-popup-deprecated.png)

### 5. "Shuffle holderB" — fixed

Follows from #3: `Shuffle what is in holderB`, `Sort what is in holderB by value`, worded like every other operation.

### 6. `holder(s)/collection` — fixed

The blank now says **`a holder or a group of them`** (and RECALL's `a deck or a group of them`). All eighteen holder parameters come from one `holderSpec()` helper now, so there is one place to word it.

### 7. Three `and`s in COUNT — fixed

`Count what is in h1 and h2, and remember it as n`, with the comma only where the widgets really are a list — `Count what is in h1 and remember it as n` is unchanged.

### 8. A widget id in `target` read as a collection — fixed

A string that names a widget of the room and no collection is now worded — and coloured — as that widget, wherever it is written: `Set the value of canvas1 to 3` under both the new and the deprecated spelling. A name no widget has stays `the canvases called myCanvases`, and a collection an earlier operation made wins over a widget of the same name, because that is the order the engine reads them in.

![routine](https://agent.virtualtabletop.io/reports/pr/1536305652446535690/ui-routine-2.png)

### 9. LABEL's noun — not changed, on purpose

LABEL is not limited to labels: it calls `setText()` on every widget it is given, so any widget with a text property is a valid target (the parameter's own comment says so, and the picker opens on all types for that reason). `the widgets called l1` is the accurate noun there, where `canvases`/`timers` are accurate for CANVAS and TIMER, which do skip widgets of the wrong type.

### 10. Width — partly, through #8

The longest new phrases were the ones #8 removed (`the canvases called canvas1` → `canvas1`), so the common case is shorter than before this PR rather than longer. The panel width itself I left alone as pre-existing.

### 11. The chip popup did not say which name it edits — fixed

`CLICK - what to look inside` for a holder against `CLICK - which widgets` for the target, and MOVE asks `where to take them from` / `where to move them to`, RECALL `where to gather them`.

![holder popup](https://agent.virtualtabletop.io/reports/pr/1536305652446535690/ui-popup-holder.png)

### 12. "deprecated" as a tooltip — fixed

`old name - click to see what to write instead`.

### 13. Link `[holder]` from the `target` lines too — done

All sixteen of them.

### 14. The tutorial paragraph

Kept, and it now names the wording it is confirming (*"which is why the routine editor words it '1 widget in each of'"*), so it reads as the explanation behind a sentence the reader has already seen rather than as a warning about something invisible.

---

**Verification.** `npm test` green (1217 tests, four new ones covering the ignored rendering, the per-holder count, the widget-id wording and the holder being offered on every operation that has one), plus `tests/testcafe/editor.js` (40 tests) and `routines.js` locally. The whole thing was walked through in a browser at 1500×1000 — the screenshots above are from that run; [here](https://agent.virtualtabletop.io/reports/pr/1536305652446535690/ui-routine-1.png) is the ignored `target` on the card itself.

VTT-AI-Agent commented:
Both findings were real and both are fixed in `58923a9`.

### 1. An explicit list of ids resolving to a same-named collection — fixed

`getCollection()` asked `collections[collection]` before it asked whether it had a list at all, so an array used as a key became the string `card` and `{"func":"DELETE","target":["card"]}` deleted the collection's widgets instead of the widget the list names. The list is checked first now, so a list of ids always means those widgets, whatever a collection happens to be called.

The ordering was already there before this PR, but it only became reachable through a normal spelling with it — a list of ids used to arrive at `getCollection()` only through `collection`/`source`/`excludeCollection`, and it is what the routine editor writes into every widgets parameter now. Thanks for catching it.

A plain **string** keeps reading as a collection first and only as a widget when no collection has that name — that is the order the engine has always used and what routines that build a collection rely on, so `target: "card"` still means the collection. The two spellings are pinned against each other in the new test.

### 2. A generated `$collection_N` overwriting a collection of the game — fixed

The counter now steps over every name already present in `collections`, so it never lands on one the game gave out (or on a temporary one of a surrounding routine that is still alive), and cleanup deletes the names this operation really created instead of counting a range — it keeps the list of them rather than recomputing it from the counter.

### Tests

`tests/client/engine/collection-names.test.js`, three cases: the list of ids beating the same-named collection, the string form still naming the collection, and a game-defined `$collection_1`…`$collection_12` surviving an operation that passes a list of ids. The first and third fail on the previous commit and pass on this one.

`npm test` green (1223 tests), `tests/testcafe/routines.js` green locally.

VTT-AI-Agent commented:
All three actionable findings were real and are fixed in `e3556a1`. Finding 4 needs no code, but I took the sentence you suggested into the PR body.

### 1. A string that names both — fixed by giving each parameter back the reading it always had

You are right that collection-first is the wrong rule for the parameters that were widget-ids only, and right that it contradicted the array rule from `58923a99`. What a plain string means is now decided by what the parameter accepted **before** this branch unified them:

| written as | reading | why |
| --- | --- | --- |
| `holder`, `fromHolder`, `toHolder`, and the old `from`/`to`/`canvas`/`label`/`timer`/`widget` | **widget first**, collection only when no widget carries the name | they named widgets and nothing else, so a game with a collection of the same name in scope keeps meaning the widget |
| `target`, `source` (SORT/SHUFFLE) | collection first, widget id as the shorthand this PR adds | they have always named a collection |
| the old `collection`, `excludeCollection`, and `source` in CLONE/SELECT/SWAPHANDS | collection and nothing else | see finding 2 |

A **list of ids is those widgets everywhere**, so the two rules never disagree about brackets — `holder: "h1"` and `holder: ["h1"]` are the same widget again.

Your repro is a test now (`ROTATE holder:'h1'` with a collection `h1` in scope rotates `c1,c2`, the children of the widget, as on `main`), together with the same collision under the old `from`/`to` spelling of MOVE.

Where a name really is ambiguous — a collection **and** a widget carry it, and the collection is not simply that one widget — the routine now reports which of the two it used, e.g. *"h1 is both a widget ID and the name of a collection - fromHolder names the widget, as it always did. Rename the collection to use it here."* The `SELECT property:'id' value:x collection:x` shape that the library games use is exempt on purpose: both readings are the same widget there, so a message would be noise.

One case your review helped me find while pinning this: `renamedParameters()` writes the current name onto the operation object, so the old spelling it came from is only visible the **first** time a routine runs. The reading is remembered per operation now, otherwise `MOVE from:'h1'` inside a FOREACH would have read `h1` one way on the first iteration and the other way on the second. That is a test too.

### 2. The old `collection` spellings accepting a bare widget id — fixed, they are strict again

`{"func":"DELETE","collection":"someWidget"}` reports *"Collection someWidget does not exist or is not an array."* and does nothing, exactly as on `main`. The same for `excludeCollection` and for `source` in CLONE/SELECT/SWAPHANDS, which never renamed and therefore never gained the shorthand. `GET collection:'auction_block'` in *Union Station* is back to the error it reports on `main` rather than reading the property off the widget.

The single-widget-id shorthand is what the **current** names offer: `DELETE target:'card1'` works, `DELETE collection:'card1'` does not. Both spellings are pinned against each other in `tests/client/engine/collection-names.test.js`.

### 3. RECALL onto a deck outside a holder — reported instead

`RECALL target:` on a deck whose parent is not a widget now pushes *"Deck dk is not in a holder, so there is nowhere to recall its cards to."* and leaves the cards where they are, instead of parenting them onto the deck. The help text says the same. A deck that does lie in a widget still recalls into it.

### 4. CALL on several widgets

No change — thanks for scanning all 573 files. The sentence about it stays in the PR body.

### Verification

`npm test` green (1228 tests; 7 new ones covering the holder/from/to collision, the second run of the same operation, the strict old spellings and RECALL on a loose deck — 5 of them fail on `58923a99`). `tests/testcafe/routines.js` and `tests/testcafe/holderevents.js` green locally. Walked in a browser at 1500×1000: with a collection `h1` in scope, `ROTATE holder:'h1'` rotates the widget's children, the routine editor says *"Rotate all widgets in h1"* with `h1` coloured as a widget rather than as a group, and the `[holder]` help topic now spells the whole order out.

The `beta` label stays off — agreed on the reasoning.
